sympy/functions/combinatorial/numbers.py

```diff
@@ -1317,65 +1333,87 @@ class genocchi(Function):
 
     .. [1] https://en.wikipedia.org/wiki/Genocchi_number
     .. [2] http://mathworld.wolfram.com/GenocchiNumber.html
+    .. [3] Peter Luschny, "An introduction to the Bernoulli function",
+           https://arxiv.org/abs/2009.06743
 
     """
 
     @classmethod
-    def eval(cls, n):
-        if n.is_Number:
-            if (not n.is_Integer) or n.is_nonpositive:
-                raise ValueError("Genocchi numbers are defined only for " +
-                                 "positive integers")
-            return 2 * (1 - S(2) ** n) * bernoulli(n)
-
-        if n.is_odd and (n - 1).is_positive:
-            return S.Zero
-
-        if (n - 1).is_zero:
-            return S.One
+    def eval(cls, n, x=None):
+        if x is S.One:
+            return cls(n)
+        elif n.is_integer is False or n.is_nonnegative is False:
+            return
+        # Genocchi numbers
+        elif x is None:
+            if n.is_odd and (n-1).is_positive:
+                return S.Zero
+            elif n.is_Number:
+                return 2 * (1-S(2)**n) * bernoulli(n)
+        # Genocchi polynomials
+        elif n.is_Number:
+            return (2 * (bernoulli(n, x) - S(2)**n * bernoulli(n, (x+S.One) / 2))).expand()
```

Comment:
This call to expand could have effects on `x` if `x` is an expression e.g.:
```python
genocchi(4, (x+y)**100)
```
Firstly only the specific type of expansion needed should be applied but also it shouldn't be applied deeply and it should only be applied to the term that needs it. Perhaps a dummy symbol should be used for the expansion and substituted with `x` after.

Reply:
You can use the specific expand function (like expand_mul or expand_multinomial) and set deep=False.

Reply:
> You can use the specific expand function (like expand_mul or expand_multinomial) and set deep=False.

I tried a few variants of that but didn't come up with something suitable in this situation (that expanded precisely the right things and not anything else). Hence my suggestion to use a dummy variable.

Reply:
Honestly this, as well as the Bernoulli polynomial, looks like something that should have special functions to generate it in the polys, similar to how the other polynomial functions work (c.f. functions in sympy/polys/orthopolys.py and the corresponding Function classes in sympy/functions/special/polynomials.py).

Reply:
> Looks like something that should have special functions to generate it in the polys

Agreed

Reply:
> I tried a few variants of that but didn't come up with something suitable in this situation (that expanded precisely the right things and not anything else). Hence my suggestion to use a dummy variable.

The dummy variable has been implemented. Moving the polynomial construction to polynomials.py I think is best left to yet another PR, where we can move Bernoulli, Genocchi and very soon Euler polynomials at once.

The three aforementioned polynomial sequences all happen to be Appell sequences, so their coefficients can be generated separately, but to do that reuqires computing the numbers individually. A recursive approach like is done for Chebyshev-T/U polynomials would be better when we effect the move.